Implementation for new Local Tunnel Debugger API

amsoedal opened the pull request:
Following up on this PR: https://github.com/Azure/vscode-kubernetes-tools-api/pull/31
New API PR: https://github.com/Azure/vscode-kubernetes-tools-api/pull/33

This PR includes the implementation for the new Local Tunnel Debugger API which adds the Debug (Local Tunnel) command and allows users to redirect traffic from their cluster to their machine when developing, allowing the benefits of remote and local debugging to be combined.

daniv-msft commented on src/api/contract/localtunneldebugger/v1.ts:
```diff
@@ -0,0 +1,15 @@
+// at all times.
+
+export interface LocalTunnelDebuggerV1 {
+    registerLocalTunnelDebugProvider(localRedirectionDebugger: LocalTunnelDebuggerV1.LocalTunnelDebugProvider): void;
```
Nit: could you please rename this one?

amsoedal replied:
Yes, thanks for catching this!

daniv-msft commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,122 @@
+        "kubernetes",
+        "contoso",
+        "megafloof",
+        "kubernetes-extension-local-tunnel-provider"
```
Should the tag be `kubernetes-extension-local-tunnel-debug-provider`? This seems to be the name used in the rest of the file.

daniv-msft replied:
...or `kubernetes-extension-local-tunnel-debugger`? Seems like we're using this name as well.
Are there any conceptual differences between both?

amsoedal replied:
Yeah, the **debug provider** would be the extension that actually provides the debugging functionality, while the **debugger** is the layer that keeps track of the debug providers and connects the Debug (Local Tunnel) command to the provider's startDebugging API. (I tried to copy the way the CloudExplorer/CloudProvider interfaces interact.) In line with this, I'll go with "kubernetes-extension-local-tunnel-debug-provider" :)

daniv-msft commented on src/components/config/config.ts:
```diff
@@ -262,3 +262,8 @@ export function isContextStatusBarDisabled(): boolean {
     return vscode.workspace.getConfiguration(EXTENSION_CONFIG_KEY)['vs-kubernetes.disable-namespace-info-status-bar'];
 }
+
+// gets the default local tunnel debug provider to use
```
Nit: Is this actually the `default`? Should it be more the `debug provider defined in settings` or something of that kind?

amsoedal replied:
Yeah, it's the one defined in settings. Will update

daniv-msft commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        let providerName: string = getLocalTunnelDebugProvider();
+        if (providerName === "") {
```
Is it possible that this `providerName` is `undefined`, if the user manually remove this field from the settings? Or do we have the guarantee that it will always be set to an empty string if missing?

amsoedal replied:
Hmm, good question. I did set the default value in package.json, and from the [docs](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration) it looks like even if they remove it from settings.json the fallback value should get resolved? However, this is an easy case to handle so I'll updating to include the `undefined` case just in case. I see some of the other configurations in this file have varying ways of handling this

daniv-msft commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,46 @@
+    startLocalTunnelDebugProvider(target?: any): void {
+        const browseExtensions = "Find Providers on Marketplace";
+        if (this.providers.length === 0)
+        {
```
Nit: could you please put this bracket on the above line, to respect the convention in the rest of the file?

amsoedal replied:
Thanks!

daniv-msft commented on src/extension.ts:
```diff
@@ -2324,7 +2328,15 @@ async function showInfoMessage(message: string) {
+
+function kubernetesFindLocalTunnelDebugProviders() {
+    searchMarketPlace("kubernetes-extension-local-tunnel-provider");
+;}
```
Nit: extra semicolon

amsoedal replied:
Thanks1

daniv-msft commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,122 @@
+# Adding Providers to the Local Tunnel Debugger
+
+The Local Tunnel Debugger is intended to provide a local debugging option that includes
+the benefits of remote debugging. Traffic from the cluster is rerouted to the version of the service 
```
Nit: maybe `application` instead of `services`? Users might not only debug services

amsoedal replied:
Sure!

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,46 @@
+import { getLocalTunnelDebugProvider } from '../config/config';
+
+export class LocalTunnelDebugger {
+    private readonly providers = Array.of<LocalTunnelDebugProvider>();
```
💡 adding some thoughts, looks like cool feature. Is there any chances of this being `undefined`? or there is a default value fallback for this?

amsoedal replied:
Hey @Tatsinnit, would you mind clarifying this one this a little more? :) I was trying to follow the way cloudexplorer keeps track of its providers, but if I'm doing something unusual here please do let me know!

Tatsinnit replied:
I think the PR for the fort below covered some parts of this: 

* Just for other readers: https://github.com/Azure/vscode-kubernetes-tools/pull/851#discussion_r552996801

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,46 @@
+            providerName = this.providers.map((p) => p.id).sort()[0];
+        }
+        const providerToUse: LocalTunnelDebugProvider | undefined = this.providers.find((p) => p.id === providerName);
+        if (providerToUse === undefined) {
```
💡 is the value of `LocalTunnelDebugProvider` always guaranteed, aka in case of empty array this `if` can be bypassed.

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,46 @@
+        this.providers.push(provider);
+    }
+
+    startLocalTunnelDebugProvider(target?: any): void {
```
💡 Idea\thought share: possible thought process behind below is: bail early if none of the conditions satisfy.

Given below 3 major if conditions, if:

* `providers.length` is `non-empty` and 
* `providename` if `non-empty` and 
*  `providerToUse` is `not undefined`, there is nothing which probably gets run?

In that case should we do all the existence check upfront and then bail with some message that you have got all necessary stuff and nothing actually happened?

Tatsinnit replied:
* Minor idea addition: we could compartmentalize these conditions into its own small `functs` units, and call it from main function as well.

amsoedal replied:
@Tatsinnit  thank you so much for adding these comments! Dan told me you had messaged him and mentioned forking this PR. Perhaps we can sync up later today? I've just addressed Dan's feedback, hopefully it won't cause any conflicts. :)

Tatsinnit replied:
Thank you so much @amsoedal, Yeah I dint wanted to just comment :) was planning to possibly add a PR to your Fork and see if that helps. Yes, please feel free to ping me internally in teams please. I think Dan will have my alias, since its git I will not ask your internal alias here.

amsoedal replied:
@Tatsinnit I merged the PR, will resolve any comments I think were addressed by it, but please feel free to reopen them or open new ones if you have more thoughts. I had to tweak a couple things (the version of tslint used in the CI build was unhappy about the use of "??", and I updated the error conditions as the latter message was always getting used). Otherwise I definitely think the code looks cleaner, really appreciate the help on that!

Tatsinnit reviewed (commented):
I added few idea shares, as this PR feature looks a cool idea. mainly all my thoughts\points below is around philosophy of possibly `succeed early if all conditions satisfy` or `bail early is some don't`

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.extension.ts:
```diff
@@ -0,0 +1,4 @@
+export interface LocalTunnelDebugProvider {
```
💡 Idea share for consistency and what others think? since this is the `interface`, we could rename this file to `localTunnelDebuggerProvider` as the interface name is **Or** rename the interface and file as `localtunneldebuggerinterface` 

* Might be worth to at least rename file same as the class name aka: rename from `localtunneldebugger.extension.ts` to `localtunneldebuggerprovider`

amsoedal replied:
That's a good point! I will rename the interface & the file to "localTunnelDebuggerProvider" and after I merge your PR. Thanks for pointing that out!

amsoedal replied:
@Tatsinnit @daniv-msft  I noticed that the "normal" debugging providers in the same repo (e.g. `dotnetdebugprovider.ts`) use this "debug provider" naming. I think "debug provider" sounds more colloquial than "debugger provider", even though the latter would technically be more consistent. But I'm open to changing it as I want to make sure I'm following the conventions of the repo & keeping things orderly!

daniv-msft replied:
Sounds good, thanks @amsoedal. Yes, if `Xdebugprovider.ts` is already in use in the codebase, I agree we should follow the existing convention and stick to "debug provider".

amsoedal replied:
Cool, thanks for the feedback. I renamed the `localtunneldebugger.extension.ts` file to `localtunneldebugprovider` like you suggested @Tatsinnit, and added the debug option to the deployments like you mentioned offline @daniv-msft. Let me know if there's more changes you guys would like to see or if it looks OK.

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,45 @@
+        }
+        const providerToUse: LocalTunnelDebugProvider | undefined = this.providers.find((p) => p.id === providerName);
+        if (providerToUse === undefined) {
+            vscode.window.showWarningMessage(`You have configured VSCode to use Local Tunnel debugger '${providerName}', but it is not installed.`, browseExtensions)
```
💡 See line 17, just wondering why this is `WarningMessage` vs `InformationMessage`? reason being this rest of the calling mechanism for both UI are same, so we can easily extract this to common static funct.

amsoedal replied:
I see you consolidated these in the PR, it looks good to me!

Tatsinnit replied:
Thanks and sounds good in that case (if things look good) you can merge my PR and then give it a test please. (It should flow through this PR automatically) :)

Tatsinnit commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,45 @@
+                    vscode.commands.executeCommand('extension.vsKubernetesFindLocalTunnelDebugProviders');
+                }
+            });
+            return;
```
💡 Idea share: this `returns undefined` but main signature returns `void` as type annotation i.e. `startLocalTunnelDebugProvider(target?: any): void `

amsoedal replied:
Ah I see -- Thanks for fixing in your PR! I want to make sure I'm following the conventions. :)

Tatsinnit commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,122 @@
+Providers must be hosted within a Visual Studio Code extension.  This table summarises what
+your providers and their hosting extension need to do; the rest of this article goes into detail.
+
+| Component            | Responsibilities                                                     |
```
💡  You can achieve this representation: 

![image](https://user-images.githubusercontent.com/6233295/104655510-ac878800-5722-11eb-9d86-1a0e19c38971.png)

From this code:

```
| Component            | Responsibilities                                                     |
|----------------------|----------------------------------------------------------------------|
| Your extension       | <ul><li>Activate when the Cluster Explorer is displayed or when the Debug (Local Tunnel) option is selected in the command palette</li><li>Register debug providers with Kubernetes extension</li><li>Tag itself as a local tunnel debug provider in the Visual Studio Marketplace</li></ul>        |
| Local Tunnel Debug provider| <ul><li>Implement the local tunnel debug provider interface</li><li>Allow users to configure which debug provider to use</li></ul>  |
| Kubernetes extension | Display the Debug (Local Tunnel) command in the palette or when interacting with supported resources in the Cluster Explorer                                 |
```

amsoedal replied:
Thanks so much! It looks so much better now :) I've updated the table items to be more descriptive as well.

Tatsinnit reviewed (commented):
Thank you so much @amsoedal Looks cool, the Final PR eye balling and key final take \ decision of how this command is hooked can be done by @itowlson , Thank you to both of you for time and letting me contribute to this PR. 🙏:)

amsoedal commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,122 @@
+### Implementing the metadata
+
+The `id` should be a user-friendly name for your debug provider, such as `Contoso Trampoline Debugger`.  
+The id will allow users to configure a default local tunnel debugger to use.
```
Good catch! Thanks!

amsoedal commented on docs/extending/localtunneldebugger.md:
````diff
@@ -0,0 +1,122 @@
+```javascript
+const MY_PROVIDER = {
+    id: "Contoso Trampoline Debugger",
+    startDebugging: (target?: any) => startDebugSession(target)
````
Oo, good question, that probably is confusing. I've added a sample implementation to localtunneldebugger.md, but should I take the extra step of renaming this method as well?

Tatsinnit replied:
💡 Sounds like a great idea. I think worth spending time to think right `funct` name to resonate right idea for end user. I will do a heavy +1 for this idea for making thinks more simpler and readable \ understandable. Thank you.

amsoedal replied:
Makes sense. I changed it to `startLocalTunnelDebugging` but I'm open to suggestions. :)

itowlson commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,143 @@
+with the Kubernetes extension. This is the responsibility of the VS Code extension that hosts
+the debug provider. To do this, the extension must:
+
+* Activate in response to the `kubernetes.cloudExplorer` view
```
I think you indicated Cluster Explorer rather than Cloud Explorer before?

In any case, is this actually required or could an extension register only when the user issues the command?  (Perhaps there are timing issues registering in response to the command - if so we should consider those.)

amsoedal replied:
Yes, you're right, thank you for catching that. Regarding the timing, I agree that it shouldn't be a requirement for the extension to activate unless the actual command is getting executed. Perhaps I can make this bullet "suggested" rather than a "must"

itowlson commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,143 @@
+* Activate in response to the `kubernetes.cloudExplorer` view
+* Activate in response to the `Debug (Local Tunnel)` command
+* Request the Kubernetes extension's Local Tunnel Debug Provider API
+* Call the `register` method for each cloud provider it wants to display
```
"cloud provider"?

amsoedal replied:
Good catch!

itowlson commented on package.json:
```diff
@@ -275,6 +277,10 @@
+                        },
+                        "vs-kubernetes.local-tunnel-debug-provider": {
+                            "type": "string",
+                            "description": "Configure the Local Tunnel debug provider you want to use by default."
```
```suggestion
                            "description": "The Local Tunnel debug provider you want to use by default."
```

itowlson commented on docs/extending/localtunneldebugger.md:
```diff
@@ -0,0 +1,143 @@
+
+    if (commandTarget) {
+        if (commandTarget.nodeType === 'resource') {
+            if (commandTarget.resourceKind.manifestKind === 'Pod') {
```
It would be good to mention in this section what kinds of resources a provider implementation should handle.  But I wonder if we need to put that burden on providers - e.g. will every provider need the same logic and UI for going from a Deployment to a Pod, and if so can we provide this in the core?

amsoedal replied:
Ooh, good point. I will definitely add which resources the provider implementation is expected to handle. Regarding whether or not to provide this in the core, let me sync offline about @daniv-msft about that, but if in practice every provider would be reimplementing the same logic to resolve the pod behind a deployment then that sounds reasonable to provide it up front.

amsoedal replied:
I synced with Dan offline, and we felt that at the moment it might not make sense to offer this in the core of the extension, as different providers could have different preferences regarding how to resolve a pod from a deployment. However this is definitely up for discussion, if you or the team have different thoughts about it!

itowlson commented on src/api/contract/localtunneldebugger/v1.ts:
```diff
@@ -0,0 +1,15 @@
+
+export interface LocalTunnelDebuggerV1 {
+    registerLocalTunnelDebugProvider(localTunnelDebugProvider: LocalTunnelDebuggerV1.LocalTunnelDebugProvider): void;
+    startLocalTunnelDebugProvider(target?: any): void;
```
Is the idea of this to allow an extension to trigger local tunnel debugging?  If so, maybe drop `Provider` from the name?

amsoedal replied:
Oh I see, e.g. "startLocalTunnelDebugging"? Yes, the idea is this actually starts the debug provider's debugging method.

amsoedal replied:
I went with "startLocalTunnelDebugSession", but I'm open to changes ofc.

itowlson replied:
If this starts the provider's session shouldn't it be a method on the provider rather than a method exposed by the Kubernetes extension?

I feel I don't have this straight in my mind yet.  Could you explain _who_ calls this method (the k8s extension, the provider, the extension containing the provider, or a third-party extension) and _when_?  Thanks and apologies for not getting it.

amsoedal replied:
No worries at all, it could definitely be a naming issue or that I'm not using the correct pattern. Basically, in extension.ts we bind the Debug (Local Tunnel) command to the method exposed by the k8s extension: 

```
registerCommand('extension.vsKubernetesDebugLocalTunnel', (target?: any) => { localTunnelDebugger.startLocalTunnelDebugSession(target); }),
```
And then inside the implementation of the method in component/localtunneldebugger.ts, we see if a provider is installed, do some tie-breaking if multiple providers are available, and then call into the provider that we chose:

```
    startLocalTunnelDebugSession(target?: any) {
        const providerSetting: string | undefined = getLocalTunnelDebugProvider();
        const providerName: string | undefined = providerSetting ? providerSetting : this.providers.map((p) => p.id).sort()[0];
        const providerToUse: LocalTunnelDebugProvider | undefined = this.providers.find((p) => p.id === providerName);

        // On success start early.
        if (providerToUse) {
            providerToUse.startLocalTunnelDebugging(target);
            return;
        }
```
Let me know if that makes sense, or if I can flatten this in a better way!

itowlson replied:
It makes sense, but that makes it sound like this is a method that is called only from _within_ the extension (from `extension.ts`), not by external consumers of the API.  In which case it should not be part of the API, just a plain exported function from the implementation module.  Unless the _provider_ extension needs to call it, there's no need to put it in that part of the API.

(More generally, unless a method gets called across extension boundaries - core calling into provider or provider calling into core - then it doesn't need to be in the API at all.  But this interface is specifically one that the provider uses to call into the core.)

Hope this makes sense - very happy to tee up a call if that would be a more efficient way to get us on the same page.

amsoedal replied:
Hey Ivan, that does make sense! Thanks for clarifying that. I don't know why I was confusing what needed to be in the API vs. what simply needs to be exported. :) I've removed the method from the interface and I tested to make sure everything was still able to snap together like it should, but don't hesitate to let me know if there's anything else I can fix, stylistically or otherwise.

itowlson commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,48 @@
+        LocalTunnelDebugger.displayUIandExecuteCommand(message);
+    };
+
+    static displayUIandExecuteCommand(message: string) {
```
Suggest e.g. `promptToFindOnMarketplace`

itowlson commented on src/components/localtunneldebugger/localtunneldebugger.ts:
```diff
@@ -0,0 +1,48 @@
+        const providerToUse: LocalTunnelDebugProvider | undefined = this.providers.find((p) => p.id === providerName);
+
+        // On success start early.
+        if (this.providers.length && providerToUse) {
```
Do you need `this.providers.length`?  If `providerToUse` is set then I think that's all we need.

amsoedal replied:
Makes sense!

itowlson commented on src/extension.ts:
```diff
@@ -2324,7 +2328,15 @@ async function showInfoMessage(message: string) {
+    searchMarketPlace("kubernetes-extension-local-tunnel-debug-provider");
+}
+
+function searchMarketPlace(searchTerm: string) {
```
super nitpick sorry
```suggestion
function searchMarketplace(searchTerm: string) {
```

amsoedal replied:
Hehe thanks for catching that!

itowlson reviewed (changes requested):
This looks terrific - thank you @amsoedal!  Most of the comments are nitpicks or typoes, but some are in docs where it would be good to clear them up.  Apart from that I think just one question about activation, another about the 'other' API function which I'm not sure of the intended use of, and an open-ended muse about whether we can/should do more setup in the core prior to calling into the provider.  But this is looking great - I'm really excited to see this in action with Bridge to Kubernetes.

amsoedal commented:
@itowlson I've addressed the feedback, thank you all again for the good comments & cooperation! If you want to see the feature in action, I'd be happy to schedule a call sometime this week and do a demo. Otherwise, please don't hesitate to let me know if there's more changes you'd like to see.

itowlson reviewed (approved):
This looks good - thanks so much for driving this through and for your patience with my niggling.

We already have a 1.2.4 release in progress, and to avoid muddying the water, I'm going to avoid merging this until after that is complete (hopefully tomorrow).  After that I'd like to do a 1.3.0 soon after (end of this week or early next week) accompanied by an updated SDK (which will also include some Configuration API changes) so that you can launch the fully integrated experience.  I'll keep you in the loop but please chase me if things are not happening as quickly as you'd like.
